File: openc3/lib/openc3/accessors/http_accessor.rb
# encoding: ascii-8bit

# Copyright 2023 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/accessors/accessor'

module OpenC3
  class HttpAccessor < Accessor
    def initialize(packet, body_accessor = 'FormAccessor', *body_accessor_args)
      super(packet)
      @args << body_accessor
      body_accessor_args.each do |arg|
        @args << arg
      end
      klass = OpenC3.require_class(body_accessor)
      @body_accessor = klass.new(packet, *body_accessor_args)
    end

    def read_item(item, buffer)
      item_name = item.name
      case item_name
      when 'HTTP_STATUS'
        return nil unless @packet.extra
        return @packet.extra['HTTP_STATUS']
      when 'HTTP_PATH'
        return nil unless @packet.extra
        return @packet.extra['HTTP_PATH']
      when 'HTTP_METHOD'
        return nil unless @packet.extra
        return @packet.extra['HTTP_METHOD']
      when 'HTTP_PACKET'
        return nil unless @packet.extra
        return @packet.extra['HTTP_PACKET']
      when 'HTTP_ERROR_PACKET'
        return nil unless @packet.extra
        return @packet.extra['HTTP_ERROR_PACKET']
      when /^HTTP_QUERY_/
        return nil unless @packet.extra
        if item.key
          query_name = item.key
        else
          query_name = item_name[11..-1]
        end
        queries = @packet.extra['HTTP_QUERIES']
        if queries
          return queries[query_name]
        else
          return nil
        end
      when /^HTTP_HEADER_/
        return nil unless @packet.extra
        if item.key
          header_name = item.key
        else
          header_name = item_name[12..-1]
        end
        headers = @packet.extra['HTTP_HEADERS']
        if headers
          return headers[header_name]
        else
          return nil
        end
      else
        return @body_accessor.read_item(item, buffer)
      end
    end

    def write_item(item, value, buffer)
      item_name = item.name
      case item_name
      when 'HTTP_STATUS'
        @packet.extra ||= {}
        @packet.extra['HTTP_STATUS'] = value.to_i
      when 'HTTP_PATH'
        @packet.extra ||= {}
        @packet.extra['HTTP_PATH'] = value.to_s
      when 'HTTP_METHOD'
        @packet.extra ||= {}
        @packet.extra['HTTP_METHOD'] = value.to_s.downcase
      when 'HTTP_PACKET'
        @packet.extra ||= {}
        @packet.extra['HTTP_PACKET'] = value.to_s.upcase
      when 'HTTP_ERROR_PACKET'
        @packet.extra ||= {}
        @packet.extra['HTTP_ERROR_PACKET'] = value.to_s.upcase
      when /^HTTP_QUERY_/
        @packet.extra ||= {}
        if item.key
          query_name = item.key
        else
          query_name = item_name[11..-1]
        end
        queries = @packet.extra['HTTP_QUERIES'] ||= {}
        queries[query_name] = value.to_s
      when /^HTTP_HEADER_/
        @packet.extra ||= {}
        if item.key
          header_name = item.key
        else
          header_name = item_name[12..-1]
        end
        headers = @packet.extra['HTTP_HEADERS'] ||= {}
        headers[header_name] = value.to_s
      else
        @body_accessor.write_item(item, value, buffer)
      end
      return value
    end

    def read_items(items, buffer)
      result = {}
      body_items = []
      items.each do |item|
        if item.name[0..4] == 'HTTP_'
          result[item.name] = read_item(item, buffer)
        else
          body_items << item
        end
      end
      body_result = @body_accessor.read_items(body_items, buffer)
      result.merge!(body_result) # Merge Body accessor read items with HTTP_ items
      return result
    end

    def write_items(items, values, buffer)
      body_items = []
      items.each_with_index do |item, index|
        if item.name[0..4] == 'HTTP_'
          write_item(item, values[index], buffer)
        else
          body_items << item
        end
      end
      @body_accessor.write_items(body_items, values, buffer)
      return values
    end

    # If this is set it will enforce that buffer data is encoded
    # in a specific encoding
    def enforce_encoding
      return @body_accessor.enforce_encoding
    end

    # This affects whether the Packet class enforces the buffer
    # length at all.  Set to false to remove any correlation between
    # buffer length and defined sizes of items in COSMOS
    def enforce_length
      return @body_accessor.enforce_length
    end

    # This sets the short_buffer_allowed flag in the Packet class
    # which allows packets that have a buffer shorter than the defined size.
    # Note that the buffer is still resized to the defined length
    def enforce_short_buffer_allowed
      return @body_accessor.enforce_short_buffer_allowed
    end

    # If this is true it will enfore that COSMOS DERIVED items must have a
    # write_conversion to be written
    def enforce_derived_write_conversion(item)
      case item.name
      when 'HTTP_STATUS', 'HTTP_PATH', 'HTTP_METHOD', 'HTTP_PACKET', 'HTTP_ERROR_PACKET', /^HTTP_QUERY_/, /^HTTP_HEADER_/
        return false
      else
        return @body_accessor.enforce_derived_write_conversion(item)
      end
    end
  end
end

File: openc3/python/openc3/accessors/http_accessor.py
# Copyright 2025 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.
#
# A portion of this file was funded by Blue Origin Enterprises, L.P.
# See https://github.com/OpenC3/cosmos/pull/1953

import re
from .accessor import Accessor
from openc3.top_level import get_class_from_module
from openc3.utilities.string import class_name_to_filename, filename_to_module, filename_to_class_name


class HttpAccessor(Accessor):
    def __init__(self, packet, body_accessor="openc3/accessors/form_accessor.py", *body_accessor_args):
        super().__init__(packet)
        self.args.append(body_accessor)
        for arg in body_accessor_args:
            self.args.append(arg)
        try:
            klass = get_class_from_module(
                filename_to_module(body_accessor),
                filename_to_class_name(body_accessor),
            )
        # Fall back to the deprecated behavior of passing the ClassName
        except ModuleNotFoundError:
            filename = class_name_to_filename(body_accessor)
            klass = get_class_from_module(f"openc3.accessors.{filename}", body_accessor)

        self.body_accessor = klass(packet, *body_accessor_args)

    def read_item(self, item, buffer):
        if item.name in [
            "HTTP_STATUS",
            "HTTP_PATH",
            "HTTP_METHOD",
            "HTTP_PACKET",
            "HTTP_ERROR_PACKET",
        ]:
            if not self.packet.extra:
                return None
            return self.packet.extra[item.name]

        item_name = item.name
        if re.match(r"^HTTP_QUERY_", item_name):
            if not self.packet.extra:
                return None
            if re.match(r"^HTTP_QUERY_", item.key):
                query_name = item_name[11:].lower()
            else:
                query_name = item.key
            queries = self.packet.extra.get("HTTP_QUERIES")
            if queries:
                return queries[query_name]
            else:
                return None

        if re.match(r"^HTTP_HEADER_", item_name):
            if not self.packet.extra:
                return None
            if re.match(r"^HTTP_HEADER_", item.key):
                header_name = item_name[12:].lower()
            else:
                header_name = item.key
            headers = self.packet.extra.get("HTTP_HEADERS")
            if headers:
                return headers[header_name]
            else:
                return None

        return self.body_accessor.read_item(item, buffer)

    def write_item(self, item, value, buffer):
        if item.name == "HTTP_STATUS":
            self.packet.extra = self.packet.extra or {}
            self.packet.extra[item.name] = int(value)
            return self.packet.extra[item.name]

        if item.name in [
            "HTTP_PATH",
            "HTTP_METHOD",
            "HTTP_PACKET",
            "HTTP_ERROR_PACKET",
        ]:
            self.packet.extra = self.packet.extra or {}
            value = str(value)
            if item.name == "HTTP_METHOD":
                value = value.lower()
            elif item.name in ["HTTP_PACKET", "HTTP_ERROR_PACKET"]:
                value = value.upper()
            self.packet.extra[item.name] = value
            return self.packet.extra[item.name]

        item_name = item.name
        if re.match(r"^HTTP_QUERY_", item_name):
            self.packet.extra = self.packet.extra or {}
            if re.match(r"^HTTP_QUERY_", item.key):
                query_name = item_name[11:].lower()
            else:
                query_name = item.key
            self.packet.extra["HTTP_QUERIES"] = self.packet.extra.get("HTTP_QUERIES", {})
            queries = self.packet.extra["HTTP_QUERIES"]
            queries[query_name] = str(value)
            return queries[query_name]

        if re.match(r"^HTTP_HEADER_", item_name):
            self.packet.extra = self.packet.extra or {}
            if re.match(r"^HTTP_HEADER_", item.key):
                header_name = item_name[12:].lower()
            else:
                header_name = item.key
            self.packet.extra["HTTP_HEADERS"] = self.packet.extra.get("HTTP_HEADERS", {})
            headers = self.packet.extra["HTTP_HEADERS"]
            headers[header_name] = str(value)
            return headers[header_name]

        self.body_accessor.write_item(item, value, buffer)
        return value

    def read_items(self, items, buffer):
        result = {}
        body_items = []
        for item in items:
            if item.name[0:4] == "HTTP_":
                result[item.name] = self.read_item(item, buffer)
            else:
                body_items.append(item)
        body_result = self.body_accessor.read_items(body_items, buffer)
        return result | body_result  # Merge Body accessor read items with HTTP_ items

    def write_items(self, items, values, buffer):
        body_items = []
        for item, index in enumerate(items):
            if item.name[0:4] == "HTTP_":
                self.write_item(item, values[index], buffer)
            else:
                body_items.append(item)
        self.body_accessor.write_items(body_items, values, buffer)
        return values

    # If this is set it will enforce that buffer data is encoded
    # in a specific encoding
    def enforce_encoding(self):
        return self.body_accessor.enforce_encoding()

    # This affects whether the Packet class enforces the buffer
    # length at all.  Set to false to remove any correlation between
    # buffer length and defined sizes of items in COSMOS
    def enforce_length(self):
        return self.body_accessor.enforce_length()

    # This sets the short_buffer_allowed flag in the Packet class
    # which allows packets that have a buffer shorter than the defined size.
    # Note that the buffer is still resized to the defined length
    def enforce_short_buffer_allowed(self):
        return self.body_accessor.enforce_short_buffer_allowed()

    # If this is true it will enforce that COSMOS DERIVED items must have a
    # write_conversion to be written
    def enforce_derived_write_conversion(self, item):
        if item.name in [
            "HTTP_STATUS",
            "HTTP_PATH",
            "HTTP_METHOD",
            "HTTP_PACKET",
            "HTTP_ERROR_PACKET",
        ]:
            return False
        if re.match(r"^HTTP_QUERY_", item.name) or re.match(r"^HTTP_HEADER_", item.name):
            return False

        return self.body_accessor.enforce_derived_write_conversion(item)
